editor/utils/with-history/index.js

```diff
@@ -14,46 +14,82 @@ import { includes } from 'lodash';
  * @return {Function} Enhanced reducer.
  */
 export default function withHistory( reducer, options = {} ) {
+	const initialPresent = reducer( undefined, {} );
+
 	const initialState = {
-		past: [],
-		present: reducer( undefined, {} ),
+		// Past already contains record of present since changes are buffered in present.
+		past: [ initialPresent ],
+		present: initialPresent,
 		future: [],
 	};
 
+	let lastAction;
+
 	return ( state = initialState, action ) => {
 		const { past, present, future } = state;
+		const previousAction = lastAction;
+		const {
+			resetTypes = [],
+			shouldOverwriteState = () => false,
+		} = options;
+
+		lastAction = action;
 
 		switch ( action.type ) {
 			case 'UNDO':
-				// Can't undo if no past
-				if ( ! past.length ) {
-					break;
+				// If there are changes in buffer, push buffer to the future.
+				if ( last( past ) !== present ) {
+					return {
+						past,
+						present: last( past ),
+						future: [ present, ...future ],
+					};
+				}
+
+				// Can't undo if no past.
+				// If the present "buffer" is the same as the last record,
+				// There is no further past.
+				if ( past.length < 2 ) {
+					return state;
 				}
 
+				const newPast = dropRight( past );
+
 				return {
-					past: past.slice( 0, past.length - 1 ),
-					present: past[ past.length - 1 ],
+					past: newPast,
+					present: last( newPast ),
 					future: [ present, ...future ],
 				};
-
 			case 'REDO':
-				// Can't redo if no future
+				// Can't redo if no future.
 				if ( ! future.length ) {
-					break;
+					return state;
+				}
+
+				return {
+					past: [ ...past, first( future ) ],
+					present: first( future ),
+					future: drop( future ),
+				};
+
+			case 'CREATE_UNDO_LEVEL':
```

Comment:
A lot of complexity in this PR is coming from the fact that we're forcing present in this "past" by this action where we could easily achieve the same behavior by just keeping the behavior of this HoR like before and just reset `lastAction` when this action is triggered which will ensure that `shouldOverwriteState` is not called on the next action and thus, a new state is always created